dir/unionDir.go

```diff
@@ -0,0 +1,114 @@
+package dir
+
+import (
+	"errors"
+	"io/fs"
+	"os"
+	"path"
+	"path/filepath"
+)
+
+// NewUnionDirFS is a virtual file system for merging multiple directories with priority
+//
+// dirs contains the union directories and the one at the forward has higher priority
+func NewUnionDirFS(dirs ...string) UnionDirFS {
+	rootedDirs := []RootedFS{}
+	for _, dir := range dirs {
+		rootedDirs = append(rootedDirs, RootedFS{os.DirFS(dir), dir})
+	}
+	return UnionDirFS{Dirs: rootedDirs}
+}
+
+// RootedFS is io.FS implementation used in New.
+// root is the root of the file system tree passed to os.DirFS.
+type RootedFS struct {
+	fs.FS
+	Root string
+}
+
+// UnionDirFS is a simple union directory file system
+//
+// it unions multiple directory to be one directory with priority based on the order of Dirs
+type UnionDirFS struct {
+	Dirs []RootedFS
+}
+
+// Open implements fs.FS interface
+//
+// traverse all union directories and return the first existing file
+func (u UnionDirFS) Open(name string) (fs.File, error) {
+	for _, dir := range u.Dirs {
+		f, err := dir.Open(name)
+		if err != nil {
+			if errors.Is(err, fs.ErrNotExist) {
+				continue
+			} else {
+				return nil, err
+			}
+		}
+		return f, nil
+	}
+	return nil, fs.ErrNotExist
+}
+
+// Path returns the path of the named file or directory under the FS
+//
+// if path exists, it returns the first existing path in union directories (dirs)
+// or if the path doesn't exist in any union directories, it return the last path of union directories(dirs)
+func (u UnionDirFS) Path(elem ...string) (string, error) {
+	var root string
+	pathSuffix := path.Join(elem...)
+	for _, dir := range u.Dirs {
+		root = dir.Root
+		file, err := dir.Open(pathSuffix)
+		if err != nil {
+			if errors.Is(err, fs.ErrNotExist) {
+				continue
+			} else {
+				return "", err
+			}
+		}
+		defer file.Close()
+		// got the first existing file path and break
+		break
+	}
+	// return the path with current OS separator
+	return filepath.Join(root, pathSuffix), nil
+}
+
+// ReadDir implements the ReadDirFile interface
+//
+// traverse all union directories and return all existing DirEntries
+func (u UnionDirFS) ReadDir(name string) ([]fs.DirEntry, error) {
+	isVisited := make(map[string]bool)
+	var newEntries []fs.DirEntry
+	for _, dir := range u.Dirs {
+		f, err := dir.Open(name)
+		if err != nil {
+			if errors.Is(err, fs.ErrNotExist) {
+				continue
+			} else {
+				return nil, err
+			}
+		}
+		defer f.Close()
+		// try to call ReadDir
+		file, ok := f.(fs.ReadDirFile)
+		if !ok {
+			return nil, &fs.PathError{Op: "readdir", Path: name, Err: errors.New("not implemented")}
+		}
+		entries, err := file.ReadDir(-1)
+		if err != nil {
+			return nil, err
+		}
+		// traverse dirEntries
+		for _, entry := range entries {
+			// skip repeated entry name
+			if _, ok := isVisited[entry.Name()]; !ok {
+				isVisited[entry.Name()] = true
```

Comment:
Yes, because this is a union directory file system. For example, /user/libexec/nottion/plugins/pluginA and /home/exampleuser/.config/notation/plugins/pluginA are valid directories to exist in the same UnionDirFS.